### PR TITLE
Updates for SketchUp 2024

### PR DESCRIPTION
**Pull request by macumber**
@jmarrec if you want to look at getting the SketchUp Plug-In working on Mac maybe we could both work on this branch

The current issue I am seeing is in `QCoreApplication* ApplicationSingleton::application(bool gui)`, `getOpenStudioModuleDirectory` is returning a bad path object.

### CODE REVIEW

**Comment by macumber** [comment]
It's loading in a debug build which doesn't help debug but does show we're not too far off

image

**Comment by macumber** [comment]
Downgrading our version of Qt to 6.5.3 to match SketchUp (6.5.2) has enabled the plug-in to load for me on Windows.

File: .github/workflows/app_build.yml [hunk 1]
```diff
@@ -15,12 +15,12 @@ env:
...
   BUILD_BENCHMARK: ON
   BUILD_PACKAGE: ON
-  QT_VERSION: 6.6.3
+  QT_VERSION: 6.5.3
 
 jobs:
   build:
-    # Note: as of 2021-01-29, this only works for push, not for pull request
-    # if: "!(contains(github.event.head_commit.message, 'skip') && contains(github.event.head_commit.message, 'ci'))"
+    if: ${{ github.event_name == 'push' || !github.event.pull_request.draft }}
```
Comment (macumber) [inline .github/workflows/app_build.yml:22]: @jmarrec what do you think about this?  I can't remember when we would want to use 'skip' or 'ci' instead?

Comment (macumber) [reply .github/workflows/app_build.yml:22]: Hmm the way I have it now marking as "Ready to Review" doesn't trigger the build, I'll work on that.

Comment (macumber) [reply .github/workflows/app_build.yml:23]: Ok now it runs when marked ready for review

**Comment by macumber** [comment]
@jmarrec I also updated to official 3.8.0 here

File: .github/workflows/app_build.yml [hunk 2]
```diff
@@ -8,19 +8,20 @@ on:
...
+    types: [ opened, reopened, synchronize, ready_for_review ]
 
 env:
   BUILD_TYPE: Release
   BUILD_DOCUMENTATION: OFF
   BUILD_TESTING: ON
   BUILD_BENCHMARK: ON
   BUILD_PACKAGE: ON
-  QT_VERSION: 6.6.3
+  QT_VERSION: 6.5.3
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply .github/workflows/app_build.yml:19]: Ok I follow now. Sketchup was using Qt 5 until 2023.1 where it switched to Qt 6.

* https://github.com/openstudiocoalition/openstudio-sketchup-plugin/issues/133#issuecomment-1871783875
* https://forums.sketchup.com/t/can-not-modify-window-style-on-sketchup-2023-1-by-win-api/241719

File: .github/workflows/app_build.yml [hunk 3]
```diff
@@ -8,19 +8,20 @@ on:
       - 'v*' # Push events to matching v*, i.e. v1.0, v20.15.10
   pull_request:
     branches: [ master, develop ]
+    types: [ opened, reopened, synchronize, ready_for_review ]
```
Comment (jmarrec) [inline .github/workflows/app_build.yml:11]: Why do you want to add this?

File: .github/workflows/app_build.yml (hunk 1 above)
Comment (jmarrec) [reply .github/workflows/app_build.yml:23]: So you can push a commit with "[skip ci]" in the commit message and the run won't be triggered.

But this is now baked into github actions anyways; https://docs.github.com/en/actions/managing-workflow-runs/skipping-workflow-runs

File: .github/workflows/app_build.yml [hunk 4]
```diff
@@ -405,7 +406,7 @@ jobs:
         cmake -E make_directory ./build
 
         if [ "$RUNNER_OS" == "macOS" ]; then
-          QT_INSTALL_DIR="/Users/irvinemac/Qt/6.6.3/macos/"
+          QT_INSTALL_DIR="/Users/irvinemac/Qt/6.5.3/macos/"
```
Comment (jmarrec) [inline .github/workflows/app_build.yml:409]:
```suggestion
          QT_INSTALL_DIR="/Users/irvinemac/Qt/$QT_VERSION/macos/"
```

**Review by jmarrec — CHANGES_REQUESTED** [review]

File: .github/workflows/app_build.yml (hunk 3 above)
Comment (macumber) [reply .github/workflows/app_build.yml:11]: They default types for pull_requests are [ opened, reopened, synchronize ].  I added ready_for_review in the case that a PR is draft and then you mark it ready for review without pushing a commit.  The idea is if you want to push code to a branch but not kick off a bunch of CI builds you can mark the PR as draft.  Then when you want CI to run you set as ready for review.  To me that seems nicer than adding [skip ci] to each commit message but we could do that too.

File: .github/workflows/app_build.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (macumber) [reply .github/workflows/app_build.yml:19]: Yeah I don't really like having our Qt version tied to SketchUp's. That is the idea of eventually removing all the Qt stuff from the SketchUp Plug-in, the biggest thing to replace is the Inspector Dialog which is not that hard but it isn't trivial.

**Comment by macumber** [comment]
I like your change @jmarrec, you good to merge this or do you want to think some more on it?

**Comment by macumber** [comment]
My guess is 15-20 hours for the inspector replacement, it would be a significant effort so I'd want to clear it with the OSC first. If it's worthwhile we can ask if that is in scope for next release.  Otherwise we can go with this approach and maybe do the html inspector later.  There will always be a binary compatibility issue with OpenStudio SDK and SketchUp due to Ruby version.

**Comment by jmarrec** [comment]
@macumber There's a decent change this will break this again:

* https://github.com/openstudiocoalition/OpenStudioApplication/issues/688


Is packaging the .so within the sketchup-plugin instead of asking for a path to the OSApp possible?

If so, perhaps we could be making the Qt version configurable, and only use 6.5.3 in specific workflows and use those binaries for Sketchup-plugin. 
Obviously that would create a hassle for maintaining (gotta compile with both versions to ensure you don't use a method that's in Qt 6.7 and not in 6.5).

File: CMakeLists.txt [hunk 5]
```diff
@@ -512,7 +512,16 @@ endif()
...
-set(QT_VERSION "6.5.3")
+set(QT_VERSION "6.5.3" CACHE STRING "Qt target version, defaults to 6.5.3")
+
+# For AboutBox, but also validates that the version is valid
+string(REGEX MATCH "^([0-9]+\\.[0-9]+)"
+  QT_VERSION_MAJOR_MINOR ${QT_VERSION})
+if(NOT QT_VERSION_MAJOR_MINOR)
+  message(FATAL_ERROR "Failed to match QT_VERSION=${QT_VERSION}")
+endif()
+set(QT_DOC_LINK "https://doc.qt.io/qt-${QT_VERSION_MAJOR_MINOR}/qtmodules.html")
```
Comment (jmarrec) [inline CMakeLists.txt:523]: Allow passing QT_VERSION... default to 6.5.3

Make a link for the aboutbox

File: ruby/openstudio_modeleditor.rb [hunk 6]
```diff
@@ -51,12 +51,12 @@ module WinAPI
...
   # if install path fails, try developer paths
-  if !File.exist?(File.join(qt_dll_path, 'Qt5Core.dll'))
+  if !File.exists?(File.join(qt_dll_path, 'Qt6Core.dll'))
     release_dll_path = File.expand_path(File.join(File.dirname(__FILE__), '../../Release/'))
     debug_dll_path = File.expand_path(File.join(File.dirname(__FILE__), '../../Debug/'))
-    if File.exist?(File.join(release_dll_path, 'Qt5Core.dll'))
+    if File.exists?(File.join(release_dll_path, 'Qt6Core.dll'))
       qt_dll_path = release_dll_path
-    elsif File.exist?(File.join(debug_dll_path, 'Qt5Core.dll'))
+    elsif File.exists?(File.join(debug_dll_path, 'Qt6Core.dll'))
```
Comment (jmarrec) [inline ruby/openstudio_modeleditor.rb:59]: Updated this

File: src/openstudio_app/AboutBox.hpp.in [hunk 7]
```diff
@@ -7,8 +7,8 @@
...
-<p>OpenStudio uses the following QT modules (version 6.3.0) that are dynamically linked using GNU Lesser General Public License (LGPL): \
-<font size='2'> Qt5Core, Qt5Widgets, Qt5Sql, Qt5Network, Qt5Xml, Qt5Concurrent, Qt5PrintSupport, Qt5Gui,Qt5Quick, Qt5QuickWidgets, Qt5Qml, Qt5WebChannel, Qt5Positioning, Qt5WebEngine, Qt5WebEngineWidgets, QtWebEngineCore, Qt5DBus, Qt5WinExtras</font> </p> \
-<p> For information on QT, please refer to <a href='https://doc.qt.io/qt-6.2/qtmodules.html'>QT</a></p> \
+<p>OpenStudio uses the following QT modules (version ${QT_VERSION}) that are dynamically linked using GNU Lesser General Public License (LGPL): \
+<font size='2'>Qt6Core, Qt6Core5Compat Qt6Widgets, Qt6Network, Qt6Concurrent, Qt6PrintSupport, Qt6Gui, Qt6Quick, Qt6QuickWidgets, Qt6Qml, Qt6QmlModels, Qt6WebChannel, Qt6Positioning, Qt6WebEngine, Qt6WebEngineWidgets, QtWebEngineCore, Qt6DBus, Qt6Xml, Qt6Svg, Qt6OpenGL</font> </p> \
+<p> For information on QT, please refer to <a href='${QT_DOC_LINK}'>QT</a></p> \
```
Comment (jmarrec) [inline src/openstudio_app/AboutBox.hpp.in:12]: Update AboutBox + dynamically set the Qt version and doc link

**Comment by jmarrec** [comment]
@bwinsatt @mioruggieroguida I'm worried that this PR will affect M2 users (#688). We do not have a M2/M3 machine for testing. Could you help us by testing the macos arm64 installer from https://github.com/openstudiocoalition/OpenStudioApplication/actions/runs/9204641443?pr=714 please?

**Comment by jmarrec** [comment]
@macumber I just realized the Tests are taking 2 hours (!!!) on mac (x64 and arm64 alike). They timeout. (And they seem to run serially and not in parallel)
Also, the benchmark step also timeouts after 4hours (when the github limit of 6hours is reached).

This is specific to this branch. Last develop build was normal

https://github.com/openstudiocoalition/OpenStudioApplication/actions/runs/9171085345/job/25214657365

**Comment by bwinsatt** [comment]
@jmarrec I can run the tests later today, but do you want me to wait until the run sequencing issue is addressed?

**Comment by jmarrec** [comment]
@bwinsatt thanks for the reply! I'd like for you to test asap, so we know whether using Qt 6.5.3 instead of 6.3.0+ breaks M2/M3 macs or not same as #688 (which we had fixed by upgrading Qt versions, and here we downgrade...)

**Comment by bwinsatt** [comment]
image

I had this error. sorry for the screenshot, I can't do anything else with this window open.

File: src/openstudio_app/AboutBox.hpp.in (hunk 7 above)
Comment (macumber) [reply src/openstudio_app/AboutBox.hpp.in:12]: Nice

**Comment by macumber** [comment]
> @macumber I just realized the Tests are taking 2 hours (!!!) on mac (x64 and arm64 alike). They timeout. (And they seem to run serially and not in parallel) Also, the benchmark step also timeouts after 4hours (when the github limit of 6hours is reached).
> 
> Example build: https://github.com/openstudiocoalition/OpenStudioApplication/actions/runs/9171085345/job/25214657365
> 
> This is specific to this branch. Last develop build was normal (4min for tests, 1min for benchmark) (cf https://github.com/openstudiocoalition/OpenStudioApplication/actions/runs/9062431730/job/24896281016)

Yikes, I tried enabling `automationmodetool	enable-automationmode-without-authentication`: https://keith.github.io/xcode-man-pages/automationmodetool.1.html

I also set Ruby 3.2.2 and Python 3.8 as the defaults on the runner.

**Comment by macumber** [comment]
I found and fixed a few bugs in the unit tests but I think the main issue is that the way we are calling processEvents without calling exec first seems to hang on Mac

https://forum.qt.io/topic/154697/processevents-blocking-on-macos-when-app-started-from-the-command-line/9
https://bugreports.qt.io/browse/QTBUG-102161

**Comment by macumber** [comment]
@bwinsatt I was able to download the package, install, and run it on a Mac M1 running Sonoma 14.4.  It seems to work, although I did see this error message related to the OpenStudio Results measure:

> Failed to infer measure name from '/Users/irvinemac/BCL/a25386cd-60e4-46bc-8b11-c755f379d916/6b37673b-b0d8-4ab4-8f5a-0a48f170b4e8/measure.rb'

**Comment by macumber** [comment]
SketchUp Plug-in is still not loading on Mac:

image

SketchUp is using Qt 6.5.2 and we are using 6.5.3, we may need to match their version exactly

**Comment by jmarrec** [comment]
> I had this error. sorry for the screenshot, I can't do anything else with this window open.

@bwinsatt unfortunately that's not our fault, it's the OS SDK, the "classic" subcommand for measure manager is broken

```
$ openstudio --version
3.8.0+f953b6fcaf
$ openstudio classic measure -s 8091
Error executing argv: ["measure", "-s", "8091"]
Error: cannot load such file -- webrick in eval:182:in `require'
```

Make sure you go to "Preferences" and uncheck the "Use Classic CLI" checkbox there. (@macumber we should probably remove it...)

**Comment by macumber** [comment]
Loading on SketchUp 2024 on Mac M1 now after matching Qt version exactly:

image

**Comment by macumber** [comment]
@bwinsatt can you try resetting your OpenStudio Application preferences manually using the workaround in https://github.com/openstudiocoalition/OpenStudioApplication/issues/717 ?  

Then after that, can you try downloading at using the package from:
https://github.com/openstudiocoalition/OpenStudioApplication/actions/runs/9258087440

Thank you so much for your help on this!

**Comment by macumber** [comment]
@jmarrec if @bwinsatt says this is good to go I think this is ready to merge.

**Comment by jmarrec** [comment]
Yeah @macumber merge at will after that.

**Comment by bwinsatt** [comment]
Can you clarify what I need to do to test on M2? I was able to install and open the application after following #717.

After opening, I get this error, but the application still is usable after clicking ok.
image

**Comment by jmarrec** [comment]
@bwinsatt That's all we needed, thanks!

that error will be resolved soon, it's an OS SDK issue.

**Pull request merged**